proof/refine/RISCV64/Invariants_H.thy

```diff
@@ -169,50 +328,117 @@ definition ntfn_q_refs_of' :: "Structures_H.ntfn \<Rightarrow> (obj_ref \<times>
+                      \<or> (\<exists>ts. ntfnObj ntfn = WaitingNtfn ts)"
+
+definition live_reply' :: "reply \<Rightarrow> bool" where
+  "live_reply' reply \<equiv> bound (replyTCB reply) \<or> bound (replyNext reply) \<or> bound (replyPrev reply)"
+
+primrec live' :: "kernel_object \<Rightarrow> bool" where
   "live' (KOTCB tcb) =
      (bound (tcbBoundNotification tcb) \<or>
-     (tcbState tcb \<noteq> Inactive \<and> tcbState tcb \<noteq> IdleThreadState) \<or>  tcbQueued tcb)"
+      bound (tcbSchedContext tcb) \<and> tcbSchedContext tcb \<noteq> Some idle_sc_ptr \<or>
+      bound (tcbYieldTo tcb) \<or>
+      tcbState tcb \<noteq> Inactive \<and> tcbState tcb \<noteq> IdleThreadState \<or>
+      tcbQueued tcb)"
+| "live' (KOCTE cte)           = False"
 | "live' (KOEndpoint ep)       = (ep \<noteq> IdleEP)"
-| "live' (KONotification ntfn) = (bound (ntfnBoundTCB ntfn) \<or> (\<exists>ts. ntfnObj ntfn = WaitingNtfn ts))"
-| "live' _                     = False"
+| "live' (KONotification ntfn) = live_ntfn' ntfn"
+| "live' (KOSchedContext sc)   = live_sc' sc"
+| "live' (KOReply r)           = live_reply' r"
+| "live' (KOUserData)          = False"
+| "live' (KOUserDataDevice)    = False"
+| "live' (KOKernelData)        = False"
+| "live' (KOArch ako)          = False"
 
 fun zobj_refs' :: "capability \<Rightarrow> obj_ref set" where
   "zobj_refs' (EndpointCap r _ _ _ _ _)  = {r}"
 | "zobj_refs' (NotificationCap r _ _ _)  = {r}"
 | "zobj_refs' (ThreadCap r)              = {r}"
-| "zobj_refs' _                          = {}"
+| "zobj_refs' (SchedContextCap r _)      = {r}"
+| "zobj_refs' (ReplyCap r _)             = {r}"
+| "zobj_refs'                 _          = {}"
 
 definition ex_nonz_cap_to' :: "obj_ref \<Rightarrow> kernel_state \<Rightarrow> bool" where
   "ex_nonz_cap_to' ref \<equiv> \<lambda>s. \<exists>cref. cte_wp_at' (\<lambda>c. ref \<in> zobj_refs' (cteCap c)) cref s"
 
 definition if_live_then_nonz_cap' :: "kernel_state \<Rightarrow> bool" where
   "if_live_then_nonz_cap' s \<equiv> \<forall>ptr. ko_wp_at' live' ptr s \<longrightarrow> ex_nonz_cap_to' ptr s"
 
-fun cte_refs' :: "capability \<Rightarrow> obj_ref \<Rightarrow> obj_ref set" where
-  "cte_refs' (CNodeCap ref bits _ _) x = (\<lambda>x. ref + (x << cteSizeBits)) ` {0 .. mask bits}"
-| "cte_refs' (ThreadCap ref) x         = (\<lambda>x. ref + x) ` dom tcb_cte_cases"
-| "cte_refs' (Zombie ref _ n) x        = (\<lambda>x. ref + (x << cteSizeBits)) ` {0 ..< of_nat n}"
-| "cte_refs' (IRQHandlerCap irq) x     = {x + (ucast irq << cteSizeBits)}"
-| "cte_refs' _ _                       = {}"
-
+primrec cte_refs' :: "capability \<Rightarrow> obj_ref \<Rightarrow> obj_ref set" where
+  "cte_refs' (UntypedCap _ _ _ _) _      = {}"
+| "cte_refs' NullCap _                   = {}"
+| "cte_refs' DomainCap _                 = {}"
+| "cte_refs' (EndpointCap _ _ _ _ _ _) _ = {}"
+| "cte_refs' (NotificationCap _ _ _ _) _ = {}"
```

Comment:
use `fun` and default case (I don't think `primrec` supports the default case syntax, otherwise primrec would be fine)